src/array/product.ts

```diff
@@ -17,13 +19,16 @@
  * // ]
  * ```
  */
-export function product<T>(arrays: T[][]): T[][] {
+export function product<T extends any[]>(
+  ...arrays: [...T]
+): Array<{ [K in keyof T]: T[K][number] }>
+export function product<T extends any[][]>(...arrays: T): T[][] {
   let out: T[][] = [[]]
   for (const array of arrays) {
-    const result: T[][] = []
+    const result = []
     for (const currentArray of out) {
       for (const item of array) {
-        const currentArrayCopy = currentArray.slice()
+        const currentArrayCopy = castArray(currentArray)
```

Comment:
Is there a reason to use castArray here?

Reply:
Yes - to perform a shallow copy of `currentArray`. I thought it'd be a good idea to use the "first-party"(to the library) utility instead of directly using `.slice()`.

Reply:
We have a policy of keeping inter-dependencies between Radashi functions to a minimum, unless maintenance burden is a concern. This helps keep each function lightweight and reduces the amount of sleuthing anyone has to do when reading the implementation of a given function. :)

Reply:
That makes a lot of sense. Thanks for bringing this to my attention, Alec. I've updated the implementation.